### PR TITLE
rfcs: new RFC on tenant capabilities

### PR DESCRIPTION
**Pull request by knz**
RFC text here: https://github.com/knz/cockroach/blob/20220810-rfc-tenant-capabilities/docs/RFCS/20220810_tenant_capabilities.md

Release note: None
Epic: CRDB-18503

### CODE REVIEW

**Review by irfansharif — COMMENTED** [review]
Should the ability to set certain zone config fields vs. not for a tenant's own data also be expressible as a "tenant capability"? (I'm thinking of https://cockroachlabs.atlassian.net/browse/CRDB-16706.)

File: docs/RFCS/20220810_tenant_capabilities.md [hunk 1]
```diff
@@ -0,0 +1,458 @@
...
+authoritative tenant's `system.settings`), span configs
+(`system.span_configurations`), etc.
+
+In a world where the tenant ID is not any more the factor that
+determines whether a tenant can "administrate the cluster", we need a
+new mechanism to point a cluster to *which tenant is currently
+authoritative for the cluster-wide system tables*.
+
+We will achieve this by introducing a new key in the system config
+span, `SystemTenantID`. Its initial value will be 1.
```
Comment (irfansharif) [inline docs/RFCS/20220810_tenant_capabilities.md:315]: How do we control which tenant is allowed to write to this key? Can there only be one?

**Review by knz — COMMENTED** [review]
> Should the ability to set certain zone config fields vs. not for a tenant's own data also be expressible as a "tenant capability"? 

Yes, that might be the right move. Can you remind us why we put this restriction in place to start with?

> Reviewed 1 of 1 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85954#-:-N9VrFjk9HEgESpQQkwI:b-8p9k4o)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian, @dt, @irfansharif, and @stevendanna)

___
*[`docs/RFCS/20220810_tenant_capabilities.md` line 315 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/85954#-N9BykZ2_8E5ECek3SlO-r1-315:-N9VrAZKDI3XwjbVwH-I:b-je8rzm) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e620390a794193d4117e3f766f95dce43784db9/docs/RFCS/20220810_tenant_capabilities.md#L315)):*
> How do we control which tenant is allowed to write to this key?

A capability :) 

Clarified in text.

**Review by irfansharif — COMMENTED** [review]
We disabled secondary tenant zone configs in 22.1 because we're using 22.1 binaries in serverless where tenants are relatively low-trust compared to other fully managed deployments. You can imagine a tenant declaring gc.ttlseconds to be arbitrarily high and accrue a ton of garbage, or set a replication factor proportional to the number of nodes in the underlying cluster, all without getting billed for it. https://github.com/cockroachdb/cockroach/issues/77344 is relevant; there are many modes through which a low-trust tenant could induce KV to do a lot of work (that issues outlines a few) so we defaulted to not allowing access to any of it. It's also not clear if serverless users will ever have access to the full gamut of zone configs (as opposed to some "fixed menu" of things). It's also not clear what things we'd like to allow in managed environments (part of why I'm asking whether they can be thought of as "capabilities" and entrusted to each cluster's operator).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85954#-:-N9X5G0K6-R1g4eul89i:bco5r89)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian, @dt, @knz, and @stevendanna)

___
*[`docs/RFCS/20220810_tenant_capabilities.md` line 315 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/85954#-N9BykZ2_8E5ECek3SlO-r1-315:-N9X3CA32FEPjjlO0n69:bgjnyaa) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e620390a794193d4117e3f766f95dce43784db9/docs/RFCS/20220810_tenant_capabilities.md#L315)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

> How do we control which tenant is allowed to write to this key?

A capability :) 

Clarified in text.
</blockquote></details>

Could you clarify if there can only be one possible writer to this state at any given point in time? This section of the proposal is gives me pause. Not around who gets to write to this key and the defacto be responsible for "hosting" cluster state, but what exactly it means for "state to be transferred across tenants". Take `system.span_configurations` for example: if another system tenant (B) were to assume ownership of this state from a previous system tenant (A), and KV points to B's `system.span_configurations`, how does the state get transferred from what A had? The level of indirection to point KV to either A's or B's state for this table is only one part of the puzzle, for such transitions each system needs to be aware of this transition to reason about the invariants they're looking to provide. For span configs specifically, if all the SQL pods observe a change in state it previously knew of (it asks KV for the contents of `system.span_configurations`) due to an ownership transition (A -> B), things will fail -- as the SQL pods continually emit "diffs" over recent changes from a view of KV state it last knew of, and SQL state it's trying to move towards (look at spanconfig.Reconciler).

Introducing this level of indirection, and possible change in state (unless we guarantee that it doesn't happen -- are we proposing mechanisms for this?), forces subsystems to have reason about this. I'm not sure if other systems are affected in this same way.

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85954)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian, @dt, @irfansharif, @knz, and @stevendanna)

___
*[`docs/RFCS/20220810_tenant_capabilities.md` line 315 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/85954#-N9BykZ2_8E5ECek3SlO-r1-315:-N9X9JcJDtQg-8kVJH7G:bmtw6a4) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e620390a794193d4117e3f766f95dce43784db9/docs/RFCS/20220810_tenant_capabilities.md#L315)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Could you clarify if there can only be one possible writer to this state at any given point in time? This section of the proposal is gives me pause. Not around who gets to write to this key and the defacto be responsible for "hosting" cluster state, but what exactly it means for "state to be transferred across tenants". Take `system.span_configurations` for example: if another system tenant (B) were to assume ownership of this state from a previous system tenant (A), and KV points to B's `system.span_configurations`, how does the state get transferred from what A had? The level of indirection to point KV to either A's or B's state for this table is only one part of the puzzle, for such transitions each system needs to be aware of this transition to reason about the invariants they're looking to provide. For span configs specifically, if all the SQL pods observe a change in state it previously knew of (it asks KV for the contents of `system.span_configurations`) due to an ownership transition (A -> B), things will fail -- as the SQL pods continually emit "diffs" over recent changes from a view of KV state it last knew of, and SQL state it's trying to move towards (look at spanconfig.Reconciler).

Introducing this level of indirection, and possible change in state (unless we guarantee that it doesn't happen -- are we proposing mechanisms for this?), forces subsystems to have reason about this. I'm not sure if other systems are affected in this same way.
</blockquote></details>

These are all great questions, which do need an answer -- but not inside the scope of this RFC.
We have a separate project post-v23.1 to organize this transition (jira: CRDB-14551) and this will deserve its own RFC.

For now, I will propagate your questions here and on the jira epic, and say we're going to find answers at the next design round.

I believe we can introduce a capability model regardless of the resolution here, simply by stating that this particular capability will need special treatment later.

**Review by irfansharif — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85954)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian, @dt, @knz, and @stevendanna)

___
*[`docs/RFCS/20220810_tenant_capabilities.md` line 315 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/85954#-N9BykZ2_8E5ECek3SlO-r1-315:-N9XOos61hpDNZOtEBWG:bhri0ne) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e620390a794193d4117e3f766f95dce43784db9/docs/RFCS/20220810_tenant_capabilities.md#L315)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

These are all great questions, which do need an answer -- but not inside the scope of this RFC.
We have a separate project post-v23.1 to organize this transition (jira: CRDB-14551) and this will deserve its own RFC.

For now, I will propagate your questions here and on the jira epic, and say we're going to find answers at the next design round.

I believe we can introduce a capability model regardless of the resolution here, simply by stating that this particular capability will need special treatment later.
</blockquote></details>

That makes sense. In the short term (23.1) it sounds like this is a particular capability ("hosts KV configuration state") that's effectively hardcoded to today's system tenant, even though we're introducing a framework to extend that to other tenants down the line (and before doing, we'd have to answer such questions). Is that a fair summary?

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85954)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian, @dt, @irfansharif, @knz, and @stevendanna)

___
*[`docs/RFCS/20220810_tenant_capabilities.md` line 315 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/85954#-N9BykZ2_8E5ECek3SlO-r1-315:-N9XV9V-7C5aXjNb1ntk:b5k2nrx) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e620390a794193d4117e3f766f95dce43784db9/docs/RFCS/20220810_tenant_capabilities.md#L315)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

That makes sense. In the short term (23.1) it sounds like this is a particular capability ("hosts KV configuration state") that's effectively hardcoded to today's system tenant, even though we're introducing a framework to extend that to other tenants down the line (and before doing, we'd have to answer such questions). Is that a fair summary?
</blockquote></details>

yes, this is a good summary. I'll reuse it in the text :)

**Review by andrewbaptist — COMMENTED** [review]
I generally agree with this thinking from a short-term migration to a unified architecture. I worry about trying to understand the burden of implementing these changes vs the benefit. My understanding is the do-nothing alternative would allow current customers to do everything they need through the system tenant already. There is likely a lot I still don't understand about how this all works today and the barriers, but I would recommend starting with a more limited set of parameters for tenants that we know they will keep long term. It is much easier to expand capabilities rather than remove them.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85954#-:-N9sJuZB7bpeHAFXWmKs:bgcbac8)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian, @dt, @irfansharif, @knz, and @stevendanna)

___
*[`docs/RFCS/20220810_tenant_capabilities.md` line 16 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85954#-N9sF-DVEE2Wnw5gFBSa:-N9sF-DVEE2Wnw5gFBSb:b-t5kz9g) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9426017b6e2a75ef40f662b9a4f4d4f7416c553/docs/RFCS/20220810_tenant_capabilities.md#L16)):*
> ```Markdown
> 
> An example capability would be the ability of a tenant to manually
> relocate ranges to particular nodes. (Currently secondary tenants do
> ```

This specific example (relocate range) seems like a bad one long-term. Relocate range can "fight" the allocator if done incorrectly, and has a high chance of causing a lot of internal work at potentially no cost (not sure how this would be billed) especially if a tenant moves ranges and then they later move back. I'm not sure this is a capability that we should ever expose to tenants. In a perfect future world, we would deprecate relocate range completely (for all cockroach users) and instead provide the capability through range policies that are also taken into account by the allocator. 
___
*[`docs/RFCS/20220810_tenant_capabilities.md` line 92 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85954#-N9sGINIC4eRqmWmeqKa:-N9sGINIC4eRqmWmeqKb:bbfghp) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9426017b6e2a75ef40f662b9a4f4d4f7416c553/docs/RFCS/20220810_tenant_capabilities.md#L92)):*
> ```Markdown
> 
>   - the ability to query cluster-wide timeseries, such as node-node
>     latencies.
> ```

Should this be done through cockroach directly long term? It seems like providing a "monitoring" service (possibly grafana) for tenants would make more sense.
___
*[`docs/RFCS/20220810_tenant_capabilities.md` line 94 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85954#-N9sGUhoDJVmRY5MO_PN:-N9sGUhoDJVmRY5MO_PO:b-y7bro2) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9426017b6e2a75ef40f662b9a4f4d4f7416c553/docs/RFCS/20220810_tenant_capabilities.md#L94)):*
> ```Markdown
>     latencies.
>   - configuring certain cluster setting that have cluster-wide impact,
>     such as the KV rebalance rate.
> ```

This is a setting that we hope to deprecate in the future. Ideally by running "as fast as possible without impacting foreground latency". There likely are other settings that will always matter here, so this is still valid. Short term I fully agree we need this, however I'm not sure that we need to expose this to tenants.
___
*[`docs/RFCS/20220810_tenant_capabilities.md` line 96 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85954#-N9sGmED5c4-hRQU2t8p:-N9sGmED5c4-hRQU2t8q:b-3r1iqi) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9426017b6e2a75ef40f662b9a4f4d4f7416c553/docs/RFCS/20220810_tenant_capabilities.md#L96)):*
> ```Markdown
>     such as the KV rebalance rate.
>   - manually placing replicas/leaseholders on specific nodes via the
>     ALTER RANGE statement.
> ```

Again something that we would like to deprecate longer term if possible, but short term we need.
___
*[`docs/RFCS/20220810_tenant_capabilities.md` line 195 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85954#-N9sIfo53zhOXbO0Vska:-N9sIfo53zhOXbO0Vskb:b-kkh10w) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9426017b6e2a75ef40f662b9a4f4d4f7416c553/docs/RFCS/20220810_tenant_capabilities.md#L195)):*
> ```Markdown
> | `RANGE_MOVE(RangeSpec...)`                | Relocate or scatter replicas or leaseholders for the specified ranges to specific nodes (see above for range spec. A tenant's own ranges are not implicitly included.)                 |
> | `RANGE_SPLIT(RangeSpec...)`               | Manual split of specified ranges (see above for range spec. A tenant's own ranges are not implicitly included.)                                                                        |
> | `SET_SYSTEM_CLUSTER_SETTING(Settings...)` | Set the specified SystemOnly cluster settings. The capability may include a range of allowed values for each setting.                                                                  |
> ```

This one in particular is problematic due to its power. There are settings that can bork a cluster if set incorrectly. However, determining valid ranges for each setting is a Herculean effort, that will almost certainly end up being too restrictive. We could take the approach that tenants have restrictions and admins don't, but then we might force the use of the system tenant more than we would like.

**Review by knz — COMMENTED** [review]
Thanks for the initial review.

> My understanding is the do-nothing alternative would allow current customers to do everything they need through the system tenant already

1) This is indeed the case, and we will recommend they do that as the default situation.
   However, we also want to bring our self-hosted customers into the multi-tenancy world, and we'd like to avoid kicking and screaming, so we also want to give them the tools to get "advanced tenants".
   Also there's another angle to look at this, outlined in the section "Advanced security: smaller-scope SRE access": we can use secondary tenants with one-off capabilities to create a sandbox for SREs to do one operation without giving them capabilities to do other things.

2) don't forget the difference between CC Dedicated and CC Serverless. In Dedicated, we want to let customers utilize their cluster the way they want, without resource constraints (they get to decide). For those, it's pretty important to let them do certain things they may have legitimate use for, such as range pre-splits and eliminate hotspots by scattering leases of pre-split ranges ahead of time.

> I would recommend starting with a more limited set of parameters for tenants that we know they will keep long term

The list in the RFC is limited to the capabilities we already know we'll need to get to feature parity on CC Dedicated.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85954#-:-N9uwVzz0M3WVU5CZrv9:b-rg68ku)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andrewbaptist, @dhartunian, @dt, @irfansharif, @knz, and @stevendanna)

___
*[`docs/RFCS/20220810_tenant_capabilities.md` line 16 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85954#-N9sF-DVEE2Wnw5gFBSa:-N9uxL6B6i69MjD2l7-k:bcm5zd2) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9426017b6e2a75ef40f662b9a4f4d4f7416c553/docs/RFCS/20220810_tenant_capabilities.md#L16)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

This specific example (relocate range) seems like a bad one long-term. Relocate range can "fight" the allocator if done incorrectly, and has a high chance of causing a lot of internal work at potentially no cost (not sure how this would be billed) especially if a tenant moves ranges and then they later move back. I'm not sure this is a capability that we should ever expose to tenants. In a perfect future world, we would deprecate relocate range completely (for all cockroach users) and instead provide the capability through range policies that are also taken into account by the allocator. 
</blockquote></details>

See my comment at top.

We can always deprecate features when we offer something better.
___
*[`docs/RFCS/20220810_tenant_capabilities.md` line 92 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85954#-N9sGINIC4eRqmWmeqKa:-N9uxPr96acBIotY7UT1:bij4y6v) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9426017b6e2a75ef40f662b9a4f4d4f7416c553/docs/RFCS/20220810_tenant_capabilities.md#L92)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

Should this be done through cockroach directly long term? It seems like providing a "monitoring" service (possibly grafana) for tenants would make more sense.
</blockquote></details>

Correct. But we are building the multi-tenancy roadmap without dependencies on other projects that have "loose" timelines, so that we can be moderately certain we can do something within a cycle.

While we're in the process of developing a new monitoring service, its delivery (especially to self-hosted customers) is still uncertain, and in the meantime we need a solution for DB Console.
___
*[`docs/RFCS/20220810_tenant_capabilities.md` line 94 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85954#-N9sGUhoDJVmRY5MO_PN:-N9uxi37CGU_iP6So_fl:bd4fxko) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9426017b6e2a75ef40f662b9a4f4d4f7416c553/docs/RFCS/20220810_tenant_capabilities.md#L94)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

This is a setting that we hope to deprecate in the future. Ideally by running "as fast as possible without impacting foreground latency". There likely are other settings that will always matter here, so this is still valid. Short term I fully agree we need this, however I'm not sure that we need to expose this to tenants.
</blockquote></details>

Let's have the mechanism, and let deployments decide whether they use it or not. They don't have to.
___
*[`docs/RFCS/20220810_tenant_capabilities.md` line 195 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85954#-N9sIfo53zhOXbO0Vska:-N9uxoTgBAhyp9HfC0Nt:b-i48zgd) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9426017b6e2a75ef40f662b9a4f4d4f7416c553/docs/RFCS/20220810_tenant_capabilities.md#L195)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

This one in particular is problematic due to its power. There are settings that can bork a cluster if set incorrectly. However, determining valid ranges for each setting is a Herculean effort, that will almost certainly end up being too restrictive. We could take the approach that tenants have restrictions and admins don't, but then we might force the use of the system tenant more than we would like. 
</blockquote></details>

This is opt-in. If not used, there's no way to customize the settings.

**Review by nvb — COMMENTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85954)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andrewbaptist, @dhartunian, @dt, @irfansharif, @knz, and @stevendanna)

___
*[`docs/RFCS/20220810_tenant_capabilities.md` line 200 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/85954#-NA5Gl4d4hop2lM7lzXl:-NA5Gl4d4hop2lM7lzXm:b4595y7) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4c53881f7e7988ba5145079410ce62d2ca26dbf/docs/RFCS/20220810_tenant_capabilities.md#L200)):*
> ```Markdown
> | `GET_SYSTEM_ZONE_CONFIG(RangeSpec...)`    | Get the zone configs for non-tenant keyspaces (range type can be "meta", "timeseries", "liveness", etc. Refers to `SHOW ZONE CONFIGURATION FOR RANGE ...`).                            |
> | `SET_SYSTEM_ZONE_CONFIG(RangeSpec...)`    | Set the zone configs for non-tenant keyspaces (see above, refers to `ALTER RANGE ... SET ZONE CONFIGURATION`).                                                                         |
> | `GET_TENANT_ZONE_CONFIG(TenantSpec)`      | Get zone configs for other tenants than self (can be specific tenant names/IDs or "all").                                                                                              |
> ```

These two capabilities feel different than the others. They're the only ones that deal with visibility into other tenants besides the current tenant. That is different than a capability that either 1) grants privileged access to a tenant's own keyspace (e.g. range splits), or 2) grants privileged access to a shared piece of infrastructure (e.g. timeseries access, liveness access, cluster-wide backup). As a result, they directly address tenancy.

Are these needed for the transparent shift from non-multitenant to single-tenant unified architecture?
___
*[`docs/RFCS/20220810_tenant_capabilities.md` line 321 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/85954#-NA5DOk48VrfRLTsfSY-:-NA5DOk48VrfRLTsfSY0:b3vfvoh) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4c53881f7e7988ba5145079410ce62d2ca26dbf/docs/RFCS/20220810_tenant_capabilities.md#L321)):*
> in the system config span

The "system config span" refers to the first 10 SQL tables in the system tenant. If we're using this `SystemTenantID` key to locate the system tenant then we have a bootstrapping problem, so this will need to be stored outside of SQL.

**Review by knz — COMMENTED** [review]
> Reviewed 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85954)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andrewbaptist, @dhartunian, @dt, @irfansharif, @knz, @nvanbenschoten, and @stevendanna)

___
*[`docs/RFCS/20220810_tenant_capabilities.md` line 200 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/85954#-NA5Gl4d4hop2lM7lzXl:-NA8h-_x04a-cWrEOq6Q:bicaj74) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4c53881f7e7988ba5145079410ce62d2ca26dbf/docs/RFCS/20220810_tenant_capabilities.md#L200)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

These two capabilities feel different than the others. They're the only ones that deal with visibility into other tenants besides the current tenant. That is different than a capability that either 1) grants privileged access to a tenant's own keyspace (e.g. range splits), or 2) grants privileged access to a shared piece of infrastructure (e.g. timeseries access, liveness access, cluster-wide backup). As a result, they directly address tenancy.

Are these needed for the transparent shift from non-multitenant to single-tenant unified architecture?
</blockquote></details>

At the time of this writing, there was a jira epic from the serverless team requesting the ability to set an arbitrary tenant's zcfgs from the system tenant. However, it's not there any more (apparently there's another way). So we can remove this capability from the proposal.
___
*[`docs/RFCS/20220810_tenant_capabilities.md` line 321 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/85954#-NA5DOk48VrfRLTsfSY-:-NA8hb6j3vlDDuQbifeB:b-gpfvxx) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4c53881f7e7988ba5145079410ce62d2ca26dbf/docs/RFCS/20220810_tenant_capabilities.md#L321)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

> in the system config span

The "system config span" refers to the first 10 SQL tables in the system tenant. If we're using this `SystemTenantID` key to locate the system tenant then we have a bootstrapping problem, so this will need to be stored outside of SQL.
</blockquote></details>

Sorry apparently I don't have the right terminology.

How do you call the span that includes /system/nodeliveness and /system/range-idgen?
that's where we need this to go.

**Review by irfansharif — APPROVED** [review]
I've had a re-read and this structure LGTM. It would be worth spelling out exactly how revocation takes effect -- we write some new proto to `system.tenants` with updated `info`, but how do stores know about it? How do they learn about updates generally -- rangefeed on that table? What guarantees are we looking to provide? But those details don't need to be fully hashed out now.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85954#-:-NDTVUDJ4yfpF48gn7D_:b-q1h3ui)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andrewbaptist, @dhartunian, @dt, @irfansharif, @knz, @nvanbenschoten, and @stevendanna)

___
*[`docs/RFCS/20220810_tenant_capabilities.md` line 200 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/85954#-NA5Gl4d4hop2lM7lzXl:-NDTWJV2AJ15XX0WwSCl:b-t7d1r7) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4c53881f7e7988ba5145079410ce62d2ca26dbf/docs/RFCS/20220810_tenant_capabilities.md#L200)):*
> there was a jira epic from the serverless team requesting the ability to set an arbitrary tenant's zcfgs from the system tenant.

I believe this was https://github.com/cockroachdb/cockroach/issues/85179, which is done.

File: docs/RFCS/20220810_tenant_capabilities.md [hunk 2]
```diff
@@ -0,0 +1,466 @@
...
+  capabilities from the tenant with ID 1.
+
+  The result is that we still have a system/admin tenant, but it's not
+  any more the tenant with ID 1.
+
+# Technical design
+
+We will introduce tenant capabilities via a new field in the `info`
+column of `system.tenants`. For this we will use richer data
+structures than simple booleans.
```
Comment (ajwerner) [inline docs/RFCS/20220810_tenant_capabilities.md:178]: Should we add a new table as opposed to a field?

File: docs/RFCS/20220810_tenant_capabilities.md [hunk 3]
```diff
@@ -0,0 +1,466 @@
...
+(alongside e.g. liveness and the range ID gen), `SystemTenantID`. Its
+initial value will be 1.
+
+Whenever we need to look up cluster-wide configuration in the KV
+layer, we will not any more use the compile-time constant 1 as the
+system tenant ID, and instead refer to that run-time configurable key.
+
+This configuration variable needs to be a system config key, and not a
+cluster setting, because we need its value before we can set up the
+cluster settings subsystem.
```
Comment (ajwerner) [inline docs/RFCS/20220810_tenant_capabilities.md:328]: What is a "system config key" and what do you mean "before we can set up the cluster settings subsytem?

File: docs/RFCS/20220810_tenant_capabilities.md [hunk 4]
```diff
@@ -0,0 +1,466 @@
...
+
+[^st]: See section below [Deciding which tenant is responsible for hosting cluster configuration](#deciding-which-tenant-is-responsible-for-hosting-cluster-configuration).
+
+We will likely also need new capabilities to issue backups for another
+tenant than the tenant itself (such that a "system tenant" can run
+BACKUP FOR TENANT on a keyspace other than its own). Once the
+capability system is in place, we can have different teams consider
+the capabilities they need separately from each other.
+
+## Mechanisms
```
Comment (ajwerner) [inline docs/RFCS/20220810_tenant_capabilities.md:212]: One detail I don't see is how tenant processes observe their own capabilities. It's assumed below but there are no specifics.

**Review by knz — COMMENTED** [review]
> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85954)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @andrewbaptist, @dhartunian, @dt, @irfansharif, @nvanbenschoten, and @stevendanna)

___
*[`docs/RFCS/20220810_tenant_capabilities.md` line 178 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/85954#-N9BykZ2_8E5ECek3SlO-r4-178:-NMAz_XJ2kSVhgZQWBWo:b7ythhb) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6f3b278d7d6d84474222019904338af1e52ab6a/docs/RFCS/20220810_tenant_capabilities.md#L178)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Should we add a new table as opposed to a field? 
</blockquote></details>

You tell me. What is your thinking here?
___
*[`docs/RFCS/20220810_tenant_capabilities.md` line 212 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/85954#-N9BykZ2_8E5ECek3SlO-r4-212:-NMAzpB4D3TMwiapiOPr:bnfj3m5) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6f3b278d7d6d84474222019904338af1e52ab6a/docs/RFCS/20220810_tenant_capabilities.md#L212)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

One detail I don't see is how tenant processes observe their own capabilities. It's assumed below but there are no specifics. 
</blockquote></details>

Correct at the time of this writing I wasn't sure this would be required. We've found out it would. It would get a new RPC on the tenant connector.
___
*[`docs/RFCS/20220810_tenant_capabilities.md` line 302 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/85954#-N9BykZ2_8E5ECek3SlO-r4-302:-NMAzvSQEs0snc9gzy3D:b-rd8i2v) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6f3b278d7d6d84474222019904338af1e52ab6a/docs/RFCS/20220810_tenant_capabilities.md#L302)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

It's not clear to me why this RFC is engaging with the problem posed in this seciton. Can you spell out why we need to couple this work to separate the "administrative tenant configuration" with any of the other work?
</blockquote></details>

I was hoping to outline a future where _all_ traffic goes through the capability system, including the definition of which tenant serves the role of system tenant. In that case, there's a chicken-and-egg problem to solve. But this work is premature and does not need to be covered here. I'll remove it.
___
*[`docs/RFCS/20220810_tenant_capabilities.md` line 328 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/85954#-N9BykZ2_8E5ECek3SlO-r4-328:-NMB-AJL-Ux0c6MDgF0S:b-ud9n0u) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6f3b278d7d6d84474222019904338af1e52ab6a/docs/RFCS/20220810_tenant_capabilities.md#L328)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

What is a "system config key" and what do you mean "before we can set up the cluster settings subsytem?
</blockquote></details>

system config key = where we store meta0 and the bootstrap keys etc.

"before we set up cluster settings" - because at this point _we don't even know which tenant keyspace to pull the cluster settings from_. 

But again this consideration belongs to a future where the choice of tenant that plays the role of "system tenant" is configurable, and so maybe it should be a different RFC.

**Review by knz — COMMENTED** [review]
I have updated the RFC to reflect the current implementation. In particular I have removed the bit that talks about routing the system administration responsibility to another tenant than the system tenant, as this will be covered by later work.

> how do stores know about it? 

We've added a cache refreshed via a rangefeed. Mentioned in RFC.

> What guarantees are we looking to provide?

Eventual consistency.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85954#-:-NU1sJEF9HBfOwtffjMu:b-fzji1b)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @andrewbaptist, @dhartunian, @dt, @irfansharif, @nvanbenschoten, and @stevendanna)

**Review by arulajmani — APPROVED** [review]

**Comment by knz** [comment]
bors r=arulajmani

**Pull request merged**